packages/actor-query-operation-bgp-left-deep-smallest-sort/lib/ActorQueryOperationBgpLeftDeepSmallestSort.ts

```diff
@@ -0,0 +1,230 @@
+   * @param {(patterns: Algebra.Pattern[]) => Promise<IActorQueryOperationOutput>} patternBinder A callback
+   * to retrieve the bindings stream of an array of patterns.
+   * @return {BindingsStream}
+   */
+  public static createLeftDeepStream(baseStream: BindingsStream, patterns: Algebra.Pattern[],
+                                     patternBinder: (patterns: Algebra.Pattern[]) =>
+                                       Promise<BindingsStream>): BindingsStream {
+    const bindingsStream: MultiTransformIterator<Bindings, Bindings> = new MultiTransformIterator(baseStream);
+    bindingsStream._createTransformer = (bindings: Bindings) => {
+      const bindingsMerger = (subBindings: Bindings) => subBindings.merge(bindings);
+      return new PromiseProxyIterator(
+        async () => (await patternBinder(ActorQueryOperationBgpLeftDeepSmallestSort.materializePatterns(patterns,
+          bindings))).map(bindingsMerger), { autoStart: true, maxBufferSize: 128 });
+    };
+    return bindingsStream;
+  }
+
+  /**
+   * Get the combined list of variables of the given pattern outputs.
+   * @param {IActorQueryOperationOutput[]} patternOutputs An array of query operation outputs
+   * @return {string[]} The array of variable names.
+   */
+  public static getCombinedVariables(patternOutputs: IActorQueryOperationOutputBindings[]): string[] {
+    return require('lodash.uniq')([].concat.apply([],
+      patternOutputs.map((patternOutput) => patternOutput.variables)));
+  }
+
+  /**
+   * Sort the given patterns and metadata by increasing estimated count.
+   * @param {IOutputMetaTuple[]} patternOutputsMeta An array of pattern output and metadata tuples.
+   * @return {IOutputMetaTuple[]} The sorted array.
+   */
+  public static sortPatterns(patternOutputsMeta: IOutputMetaTuple[]): IOutputMetaTuple[] {
+    return require('lodash.sortby')(patternOutputsMeta, [(e: IOutputMetaTuple) => {
+      return ActorQueryOperationBgpLeftDeepSmallestSort.getTotalItems(e.meta);
+    }]);
+  }
+
+  /**
+   * Estimate an upper bound for the total number of items from the given metadata.
+   * @param {{[p: string]: any}} smallestPattern The optional metadata for the pattern
+   *                                             with the smallest number of elements.
+   * @param {{[p: string]: any}[]} otherPatterns The array of optional metadata for the other patterns.
+   * @return {number} The estimated number of total items.
+   */
+  public static estimateCombinedTotalItems(smallestPattern: {[id: string]: any},
+                                           otherPatterns: {[id: string]: any}[]): number {
+    const smallestCount: number = ActorQueryOperationBgpLeftDeepSmallestSort.getTotalItems(smallestPattern);
+    return otherPatterns
+      .map((otherPattern) => smallestCount * ActorQueryOperationBgpLeftDeepSmallestSort.getTotalItems(
+        otherPattern))
+      .reduce((sum, element) => sum + element, 0);
```

Comment:
This reminds me I also saw this in the other actor. What is the reason for using a sum here?

Reply:
It's just a heuristic to estimate the upper bound of the total number of items for the whole BGP.